Write splitting BAM indexes, and code for merging sharded BAM or CRAM files

tomwhite opened the pull request:
This moves code from GATK to Hadoop-BAM for merging sharded BAM or CRAM files. It also adds an option to write splitting BAM indexes from MR (defaults to off). The splitting indexes (if present) are merged into a single index as a part of the merge operation.

tomwhite commented:
Addresses #48, #60, and #61.

fnothaft commented:
+1, we have similar code in ADAM, will be good to eliminate it.

tomwhite commented:
@fnothaft would you (or someone else on the ADAM team) be able to review this? I tried running GATK with this change and it passes tests - would be good for someone to do the same with ADAM.

fnothaft commented:
@tomwhite I'm traveling this week so I may be a bit slow, but I should be able to make a pass by the end of the week. Would that timing be OK for you?

tomwhite commented:
@fnothaft, that would be great - thanks!

tomwhite commented:
I've updated this so that it uses the Java NIO file API for metadata and merging operations (this allows it to be used from GATK which supports other NIO implementations). I've also added merge for VCF files.

@fnothaft can you take a look please? @cmnbroad, it would be great to get your feedback too.

cmnbroad commented on src/main/java/org/seqdoop/hadoop_bam/util/SAMFileMerger.java:
```diff
@@ -0,0 +1,150 @@
+          "files were found in " + partPath);
+    }
+
+    Files.deleteIfExists(outputPath);
```
Should we be deleting the index here as well ?

tomwhite replied:
Yes. Fixed.

cmnbroad commented on src/main/java/org/seqdoop/hadoop_bam/util/NIOFileUtil.java:
```diff
@@ -0,0 +1,88 @@
+      Files.copy(part, out);
+    }
+    for (final Path part : parts) {
+      Files.delete(part);
```
It looks like (both) mergeParts implementations delete these at the end as well, so I'm wondering if we need this here. Doing it at the end after everything succeeds makes sense to me.

tomwhite replied:
This method deletes the part files after they have been merged, so this makes sense here. The directory that contained the part files (and also metadata like _SUCCESS, and index files) is deleted after the whole operation succeeded.

cmnbroad commented on src/main/java/org/seqdoop/hadoop_bam/util/SAMFileMerger.java:
```diff
@@ -0,0 +1,150 @@
+    try (final OutputStream out = Files.newOutputStream(outputSplittingBaiPath)) {
+      mergeSplittingBaiFiles(out, partPath, headerLength, fileLength);
+    } catch (IOException e) {
+      deleteRecursive(outputSplittingBaiPath);
```
I think this should rethrow, otherwise we're masking this failure.

tomwhite replied:
Done.

cmnbroad commented on src/main/java/org/seqdoop/hadoop_bam/util/SAMFileMerger.java:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    List<Path> parts = getFilesMatching(partPath,
+        "glob:**/part-[mr]-[0-9][0-9][0-9][0-9][0-9]*",
```
This glob string should be a constant somewhere as its repeated in several places.

tomwhite replied:
Done

cmnbroad commented on src/main/java/org/seqdoop/hadoop_bam/SplittingBAMIndexer.java:
```diff
@@ -194,6 +194,13 @@ public void processAlignment(final SAMRecord rec) throws IOException {
 		count++;
 	}
 
+	void processAlignment(final long virtualOffset) throws IOException {
```
This doesn't appear to be used anywhere?

tomwhite replied:
It's used in BAMRecordWriter to write splitting BAM indexes from MR. See https://github.com/HadoopGenomics/Hadoop-BAM/pull/111/commits/cd89366d86302ba7d0c02079fc3d663dd46408ce#diff-8ce0e02bf18a51ccfbbe62aebc2f9498R141.

cmnbroad commented on src/main/java/org/seqdoop/hadoop_bam/util/VCFFileMerger.java:
```diff
@@ -0,0 +1,127 @@
+   * @param header the header for the merged file
+   * @throws IOException
+   */
+  public static void mergeParts(final String partDirectory, final String outputFile,
```
This fails with .bcf files (even a legitimate .bcf that htsjdk can handle, and that passes the testRoundTrip test, fails the testRoundTripWithMerge test). I'm guessing its not a quick fix though, so we might want to make a separate ticket, and for now just reject them here and document the restriction.

tomwhite replied:
I've added a check so that BCF files are rejected.

tomwhite commented:
Thanks for the review @cmnbroad! I've addressed your feedback in a new commit.

cmnbroad commented:
LGTM.
